File: Javascript/Behavioral Patterns/Observer/observer.ts
interface Observer {
    update(news: string): void
}

interface Subject {
    attach(observer: Observer): void
    detach(observer: Observer): void
    notify(): void
}

export class NewsAgency implements Subject {
    private observers: Observer[] = []
    private latesNews: string = ""
    
    attach(observer: Observer): void {
        this.observers.push(observer)
    }

    detach(observer: Observer): void {
        this.observers = this.observers.filter(obs => obs !== observer)
    }

    setNews(news: string): void {
        this.latesNews = news
        this.notify()
    }

    notify(): void {
        this.observers.forEach(observer => observer.update(this.latesNews))
    }
}

export class Subscriber implements Observer {
    private name: string

    constructor(name: string) {
        this.name = name
    }

    update(news: string): void {
       console.log(`${this.name} received latest new: ${news}`) 
    }
}

File: Go/Behavioral_Patterns/Observer/observer.go
package observer

import (
	"fmt"
	"slices"
)

type Observer interface {
	Update(news string)
}

type Subject interface {
	Attach(observer Observer)
	Detach(observer Observer)
	SetNews(new string)
	Notify()
}

type Agency struct {
	observers []Observer
	latesNews string
}

func (a *Agency) Attach(obs Observer) {
	a.observers = append(a.observers, obs)
}

func (a *Agency) Detach(obs Observer) {
	for i, observer := range a.observers {
		if observer == obs {
			a.observers = slices.Delete(a.observers, i, i+1)
			break
		}
	}
}

func (a *Agency) SetNews(new string) {
	a.latesNews = new
}

func (a *Agency) Notify() {
	for _, observer := range a.observers {
		observer.Update(a.latesNews)
	}
}

func NewAgency() *Agency {
	return &Agency{
		observers: []Observer{},
		latesNews: "",
	}
}

type Subscriber struct {
	name string
}

func (s *Subscriber) Update(news string) {
	fmt.Printf("%s received latest new: %s\n", s.name, news)
}

func NewSubscribe(name string) *Subscriber {
	return &Subscriber{
		name: name,
	}
}
